reproduce-figures/utils.R

```diff
@@ -75,3 +72,123 @@ prepare_gene_expression_tsv <- function(sce, marker_genes, output_tsv) {
+#' @param output_tsv Path to output TSV file
+#'
+prepare_sample_metadata <- function(
+    portal_metadata, 
+    output_tsv) {
+  
+  sample_metadata <- portal_metadata |>
+    dplyr::select(scpca_project_id, scpca_sample_id, diagnosis, disease_timing, is_cell_line) |>
+    # remove duplicate rows, which occur when there are multiple libraries per sample
+    dplyr::distinct()
+  
+  # Prepare data frame with samples whose full metadata is not yet available on the portal
+  bulk_only_sample_metadata <-  tibble::tribble(
+    ~scpca_project_id, ~scpca_sample_id, ~diagnosis,        ~disease_timing,     ~is_cell_line, 
+    ###########################################################################################
+    "SCPCP000006",     "SCPCS000210",    "Wilms tumor",     "Initial diagnosis", FALSE,
+    "SCPCP000006",     "SCPCS000211",    "Wilms tumor",     "Initial diagnosis", FALSE,
+    "SCPCP000009",     "SCPCS000129",    "Medulloblastoma", "Initial diagnosis", FALSE,
+    "SCPCP000017",     "SCPCS000606",    "Osteosarcoma",    "Recurrence",        FALSE
+  )
+  
+  sample_metadata |>
+    dplyr::bind_rows(bulk_only_sample_metadata) |>
+    readr::write_tsv(output_tsv)
+}
+  
+  
+  
+
+
+#' Prepare and export library metadata file
+#'
+#' @param portal_metadata Data frame of portal-wide metadata
+#' @param bulk_metadata_dir Directory with bulk metadata TSV files
+#' @param project_metadata_dir Directory with project-specific metadata TSV files
+#' @param output_tsv Path to output TSV file
+#'
+prepare_library_metadata <- function(
+    portal_metadata, 
+    bulk_metadata_dir, 
+    project_metadata_dir, 
+    output_tsv) {
+  
+  # final columns to include in the exported library metadata tsv
+  final_columns <- c(
+    "scpca_project_id",
+    "scpca_sample_id",
+    "scpca_library_id",
+    "seq_unit",
+    "technology"
+  )
+  
+  library_metadata <- portal_metadata |>
+    # Group multiplexed sample ids back together; keep has_cellhash for later manipulation
+    dplyr::group_by(scpca_project_id, scpca_library_id, seq_unit, technology, has_cellhash) |>
+    dplyr::summarize(scpca_sample_id = paste(scpca_sample_id, collapse = ";")) |>
+    dplyr::ungroup()
+  
+  # duplicate multiplexed rows so we have a row for cellhash technology
+  cellhash_rows <- library_metadata |>
+    dplyr::filter(has_cellhash) |>
+    # in our code, we detect this technology with "cellhash" only, so the version isn't needed
+    dplyr::mutate(technology = "cellhash")
+    
+  library_metadata <- library_metadata |>
+    dplyr::bind_rows(cellhash_rows) |>
+    dplyr::select(-has_cellhash)
+  
+  # Define bulk and project metadata files
+  bulk_files <- list.files(
+    path = bulk_metadata_dir, 
+    pattern = "*_bulk_metadata\\.tsv", 
+    full.names = TRUE
+  )
+  citeseq_files <- list.files(
+    path = project_metadata_dir, 
+    pattern = "*_single_cell_metadata\\.tsv", 
+    full.names = TRUE
+  ) 
```

Comment:
I think same comment here. You can just provide one argument `project_scratch_dir` and grab the metadata files from there by recursively searching and then filtering based on the project ID being a CITE seq project.